error.go

```diff
@@ -204,8 +205,8 @@ func newErrMissingType(c *Container, k key) errMissingType {
 
 	err := errMissingType{Key: k}
 	for _, t := range suggestions {
-		k.t = t
-		if len(c.providers[k]) > 0 {
+		if len(c.getValueProviders(k.name, t)) > 0 {
+			k.t = t
```

Comment:
Mutating `k.t` before appending to suggestions isn't ideal. It might be worth adding methods to `errMissingType` with a similar set signature as `getValueProviders` above so that this call is a little more explicit. This obviously doesn't need to happen here, but I think it's worth noting for a potential future change.

Reply:
Fair. This will change a bit in a follow-up PR and I'll see if the
`k.t` mutation can be simplified.